[Accessibility] Improve checkout error messages

Manussakis opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This is the second PR dedicated to close the #37602 issue.
**First PR**: https://github.com/woocommerce/woocommerce/pull/48738

Closes #37602  .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR improves the form error messages on the classic checkout page.

1. Add a product to the cart.
2. Go to the classic checkout page.
3. Turn on the screen reader.
4. Submit the form with the required fields empty.
6. Verify a notice element appears at the top of the form. The notice element style may vary per theme.
![image](https://github.com/woocommerce/woocommerce/assets/9420947/ab723a51-84b8-4126-aa68-b597ce1130e0)
7. Verify the notice element contains an error message for each one of the required fields with error.
8. Verify the focus went to the notice element and the screen reader announced its content.
9. Verify you can tab through the error messages inside the notice element.
10. The hover style of the links inside the notice element depends on the theme installed.
11. Verify each one of the error messages moves the focus to the associated form field once clicked.
14. Verify the fields with errors have an inline message below them.
15. By default, the style of the inline message is: `color: #a00;`,  `font-size: .75em;`, and `margin-bottom: 0;`. However, it can vary depending on the installed theme.
![image](https://github.com/woocommerce/woocommerce/assets/9420947/8908cb9c-6ae7-46e5-a4f1-23096d649188)
16. The inline messages should disappear once the field value is updated and not appear again until the form has been submitted. This behavior is because the error messages come translated from the server and can be different depending on the value of the input.
17. Perform the steps above logged in/out.
18. Perform the test above with the "Shipping to a different address" form opened.
19. The logic that validates the inputs on the client side when their value change remains the same.
20. Perform the tests against the WP default themes:
    - Twenty Seventeen
    - Twenty Nineteen
    - Twenty Twenty
    - Twenty Twenty-One
    - Twenty Twenty-Two
    - Twenty Twenty-Three
    - Twenty Twenty-Four

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Manussakis committed:
Improve required info for screen readers on checkout fields

Manussakis committed:
Add style to inline error message on checkout fields

Manussakis committed:
Add id attribute to terms error

Manussakis committed:
Improve accessibility of checkout errors messages

Manussakis committed:
Prevent screen readers from reading the asterisk on checkout fields

Manussakis committed:
Revert spacing change

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/address-i18n.js:
```diff
@@ -17,7 +17,10 @@ jQuery( function( $ ) {
 					'&nbsp;<abbr class="required" title="' +
 					wc_address_i18n_params.i18n_required_text +
-					'">*</abbr>'
+					'" aria-hidden="true">*</abbr>' + 
```
Prevent screen readers from announcing the asterisk and make them announce the term "required" instead.

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/address-i18n.js:
```diff
@@ -17,7 +17,10 @@ jQuery( function( $ ) {
 					'&nbsp;<abbr class="required" title="' +
 					wc_address_i18n_params.i18n_required_text +
-					'">*</abbr>'
+					'" aria-hidden="true">*</abbr>' + 
+					'<span class="screen-reader-text">' + 
+					wc_address_i18n_params.i18n_required_text +
+					'</span>'
```
Prevent screen readers from announcing the asterisk and make them announce the term "required" instead.

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/checkout.js:
```diff
@@ -35,7 +35,7 @@ jQuery( function( $ ) {
 
 			// Inline validation
-			this.$checkout_form.on( 'input validate change', '.input-text, select, input:checkbox', this.validate_field );
+			this.$checkout_form.on( 'input validate change focusout', '.input-text, select, input:checkbox', this.validate_field );
```
Run the `validate_field` callback when the focus moves away from the input.

Manussakis committed:
Revert spacing change

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/checkout.js:
```diff
@@ -209,16 +209,16 @@ jQuery( function( $ ) {
-						$parent.removeClass( 'woocommerce-validated' ).addClass( 'woocommerce-invalid woocommerce-invalid-required-field' );
-						validated = false;
-					} else if ( $this.val() === '' ) {
+					if ( ( 'checkbox' === $this.attr( 'type' ) && ! $this.is( ':checked' ) ) || $this.val() === '' ) {
```
Small refactor – replace the `else if` statement with `||`.

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/checkout.js:
```diff
@@ -550,7 +556,14 @@ jQuery( function( $ ) {
 							// Add new errors
 							if ( result.messages ) {
-								wc_checkout_form.submit_error( result.messages );
+								var $msgs = $( result.messages )
+									.removeAttr( 'role' )
+									.attr( 'tabindex', '-1' );
```
1. Overall, `<ul>` elements must have the default `list` role (`role="list"`). This became a problem in this task because the list items have been wrapped by the `<a>` tag and Voice Over keeps announcing the content infinitely.
The solution was to remove the `role="alert"` attribute from the `<ul>` element and add it to the parent.

2. The `tabindex="-1"` attribute is needed so the live region can be focused once populated. This allows users to tab through the links inside it and go straight to the input they want to fix.

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/checkout.js:
```diff
@@ -550,7 +556,14 @@ jQuery( function( $ ) {
+								var $msgsWithLink = wc_checkout_form.wrapMessagesInsideLink( $msgs );
+								var $msgsWrapper = $( '<div role="alert"></div>' ).append( $msgsWithLink );
+
+								wc_checkout_form.submit_error( $msgsWrapper.prop( 'outerHTML' ) );
```
The `submit_error` method expects the message content as a string. The conversion is done by calling `prop( 'outerHTML' )`.

Manussakis commented on plugins/woocommerce/includes/class-wc-checkout.php:
```diff
@@ -924,7 +924,7 @@ protected function validate_checkout( &$data, &$errors ) {
 		// phpcs:ignore WordPress.Security.NonceVerification.Missing
 		if ( empty( $data['woocommerce_checkout_update_totals'] ) && empty( $data['terms'] ) && ! empty( $data['terms-field'] ) ) {
-			$errors->add( 'terms', __( 'Please read and accept the terms and conditions to proceed with your order.', 'woocommerce' ) );
+			$errors->add( 'terms', __( 'Please read and accept the terms and conditions to proceed with your order.', 'woocommerce' ), array( 'id' => 'terms') );
```
Add the `id` attribute to the "terms" error message so it can be associated with the input.

Manussakis commented on plugins/woocommerce/includes/wc-template-functions.php:
```diff
@@ -2854,7 +2854,8 @@ function woocommerce_form_field( $key, $args, $value = null ) {
 		if ( $args['required'] ) {
 			$args['class'][] = 'validate-required';
-			$required        = '&nbsp;<abbr class="required" title="' . esc_attr__( 'required', 'woocommerce' ) . '">*</abbr>';
+			$required        = '&nbsp;<abbr class="required" title="' . esc_attr__( 'required', 'woocommerce' ) . '" aria-hidden="true">*</abbr><span class="screen-reader-text">' . esc_html__( "Required", "woocommerce" ) . '</span>';
+			$args['custom_attributes']['aria-required'] = 'true';
```
Prevent screen readers from announcing the asterisk and make them announce the term "required" instead.

Manussakis committed:
Add changelog file

Manussakis committed:
Decrease line length

Manussakis committed:
Lowercase the required term

Manussakis committed:
Fix query methods in checkout tests

Manussakis committed:
Add space before required text for screen readers

Manussakis committed:
Fix query of shipping fields on legacy checkout tests

Manussakis committed:
Remove asterisk from field name on legacy checkout test

Manussakis committed:
Remove invalid character in Phone and Email fields on legacy checkout test

Manussakis committed:
Add asterisk to get phone and email by label on legacy checkout test

Manussakis committed:
Fix field labels for legacy checkout tests

Manussakis commented on plugins/woocommerce/tests/e2e-pw/tests/shopper/cart-checkout-calculate-tax.spec.js:
```diff
@@ -267,13 +271,13 @@ test.describe.serial(
 						.type( customer.billing.us.zip );
 					await page
-						.getByLabel( 'Phone *' )
+						.getByLabel( 'Phone * required' )
```
The `getByLabel` method includes the asterisk even if it has the `aria-hidden="true"` attribute. That's why it hasn't been removed in this case.

Manussakis commented:
@dkotter This is the second PR dedicated to closing the https://github.com/woocommerce/woocommerce/issues/37602 issue. It's ready for code review.

-----

For the record, this is the first PR: https://github.com/woocommerce/woocommerce/pull/48738

dkotter reviewed (approved):
@ankitguptaindia @qasumitbagthariya This is ready for QA

qasumitbagthariya commented:
### QA Update ✅ 

-------------

 I have checked this issue in the `Manussakis:fix/37602-checkout-form-error-feedback` branch working as expected.


**Testing Environment** 

<Details>

- WordPress: 6.5.5
- Theme: Twenty Twenty-Four 1.1
- WooCommerce - 9.0.2
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 

</Details>

Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)


https://github.com/woocommerce/woocommerce/assets/67687255/120e7764-584c-49a3-b028-afcef68f8639

vikrampm1 commented:
@Manussakis please remove [WIP] and make this PR ready for review.

@coreymckrill this PR is ready for Woo review.

Manussakis committed:
Merge branch 'woocommerce:trunk' into fix/37602-checkout-form-error-feedback

senadir reviewed (approved):
Thank you for working on this, I followed the testing steps logged in and not, as well as reviewing the code.

senadir committed:
Merge branch 'trunk' into fix/37602-checkout-form-error-feedback

Manussakis committed:
Fix php lint errors

Manussakis committed:
Merge branch 'trunk' into fix/37602-checkout-form-error-feedback

Manussakis committed:
Add required tem to the Government ID input on tests

Manussakis committed:
Merge branch 'trunk' into fix/37602-checkout-form-error-feedback

Manussakis committed:
Merge branch 'trunk' into fix/37602-checkout-form-error-feedback

jeffpaul commented:
@senadir is there anything additional needed from us before this is good to merge in / prepare for a Woo release?

vikrampm1 commented:
@senadir this PR has completed all the steps from our end and is ready to be merged by you / Woo.

opr commented:
@vikrampm1 this PR has merge conflicts. I'd recommend rebasing on WooCommerce trunk and we will re-run CI.

Manussakis committed:
Merge branch 'trunk' into fix/37602-checkout-form-error-feedback

Manussakis commented:
Several CI tests failed after merging trunk into this branch.
I will work on fixing it and leave a comment when it's ready for review again.

cc: @vikrampm1 .

Manussakis committed:
Revert changes on required fields label

Manussakis commented:
I have removed the code that hides for screen readers the asterisk of the required input labels. 
[This PR](https://github.com/woocommerce/woocommerce/pull/49696) is already handling this issue. Therefore, I removed the following step from the QA instruction: 
> 12. Verify the asterisk is not announced by the screen reader when the field is focused ("Star").

@opr I have rebased on WooCommerce trunk.

Manussakis commented:
@opr Please hold off on merging this PR. I realized it has an issue similar to [the one mentioned in this PR](https://github.com/woocommerce/woocommerce/pull/48738#pullrequestreview-2204533147).

I will add styles for the input inline error message to the default WP themes.

Manussakis committed:
Create checkout-inline-error-message mixin

Manussakis committed:
Replace SCSS variable with a CSS one

Manussakis committed:
Add checkout inline error message to T17

Manussakis committed:
Add checkout inline error message to T19

Manussakis committed:
Add checkout inline error message to TT

Manussakis committed:
Add checkout inline error message to TT1

Manussakis committed:
Add checkout inline error message to TT2

Manussakis committed:
Add checkout inline error message to TT3

Manussakis committed:
Include notice banner block in the notice selectors

Manussakis committed:
Add inline documentation to role attribute removal

Manussakis commented:
@dkotter This PR has undergone some changes.
Please check these two comments for more context:
- https://github.com/woocommerce/woocommerce/pull/49094#issuecomment-2313338965
- https://github.com/woocommerce/woocommerce/pull/49094#issuecomment-2313641067

Please re-review it. Thanks!

Manussakis commented:
@qasumitbagthariya Please, test this PR again. Thanks!

qasumitbagthariya commented:
### QA Update ✅ 

-------------
@Manussakis 

 I have checked this issue in the `Manussakis:fix/37602-checkout-form-error-feedback` branch working as expected.

Cross check with different theme like

1. Storefront
2. TT4

**Testing Environment** 

<Details>

- WordPress: 6.5.5
- Theme: Twenty Twenty-Four 1.1
- WooCommerce - 9.4.0-dev
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 

</Details>

https://github.com/user-attachments/assets/de08a3a4-47d4-45fd-b02c-281395d7ff20


https://github.com/user-attachments/assets/2bbb3196-4d2a-451c-be40-d0ecf7b55947

Update:
@vikrampm1 PR is ready for the next step.

vikrampm1 commented:
@opr this PR is ready for your re-review and then merging, once it is good to go.

jeffpaul commented:
@senadir @opr are there any additional code review concerns here or is this ready to merge?

opr commented:
Hi - I re-ran CI and all the jobs are passing besides ones that can't because it's a fork.

@senadir since you and @dkotter have approved this I will go ahead and merge.

senadir commented:
Looking into it, I missed several pings here.

senadir reviewed (approved):
Thank you! I tested this with a variety of themes and it's looking good.

I do have one question, when errors are not matching, only one of them is visible.

For example, when the form is empty, those are the visible errors:
<img width="817" alt="image" src="https://github.com/user-attachments/assets/435b930d-71ae-4840-9b1b-39bf0710fd72">


If I insert a wrong email (`test.test`), those errors are no longer shown in the notice:
<img width="845" alt="image" src="https://github.com/user-attachments/assets/e1085b58-3aad-4ed2-a3c1-81595d4569f2">


Previously visible errors messages are gone, but the field still has the red border.

Would that be outside scope for this PR?

Manussakis commented:
@senadir Thanks for testing it.

This PR takes the checkout error messages and makes their UI accessible, but the logic to determine what the error messages should look like is out of its scope.

I followed the steps you listed to reproduce the issue on the trunk branch. As you can see in the screenshot, even having other fields with errors, only the email error message is returned, just like you flagged.

![Screenshot 2024-10-03 at 19 03 44](https://github.com/user-attachments/assets/73432c84-2ff3-41ad-bc6c-8835af2a1964)

Thanks!

vikrampm1 commented:
@senadir please review the above comment and proceed with merging the PR if everything looks good.

vikrampm1 commented:
@senadir just a reminder on the above.

senadir commented:
Thank you! good with me for now.
